### PR TITLE
fix(hpa): remove problematic explicit setting of replica counts in deployments

### PR DESCRIPTION
**Pull request by samcgardner**
This change fixes https://github.com/ory/k8s/issues/318 

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
      and signed the CLA.
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added necessary documentation within the code base (if
      appropriate).

### CODE REVIEW

File: helm/charts/kratos/templates/hpa.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{{- if .Values.autoscaling.enabled }}
+apiVersion: autoscaling/v2beta1
```
Comment (piotrmsc) [inline helm/charts/kratos/templates/hpa.yaml:2]: nit: In hydra we already use beta2
```suggestion
apiVersion: autoscaling/v2beta2
```

File: helm/charts/kratos/values.yaml [hunk 2]
```diff
@@ -267,6 +267,14 @@ securityContext:
   allowPrivilegeEscalation: false
   privileged: false
 
+# Horizontal pod autoscaling configuration
+autoscaling:
```
Comment (piotrmsc) [inline helm/charts/kratos/values.yaml:271]: In the issue you've mentioned kratos is not impacted however I see you did the same change as in hydra?
Also with hydra by default we give HPA with max to 3 replicas (we don't have demo profiles etc) so maybe let's limit a bit 100 replicas ;-)

**Comment by piotrmsc** [comment]
Thank you for your contribution :) Does it work for you in your git ops env? I'm asking because 
with this change, after templating charts with autoscaling enabled deployment won't have replicas set, HPA will set it but then if this deployment is watched by the controller it will notice that `replicas` property has been and will try to remove it as part of reverting to the original state.  Or it ignores new fields added to the spec?

What tools do you use ?:) I assume you have gitOps controller + some helm controller to deploy stuff picked up by git controller?

File: helm/charts/kratos/values.yaml (hunk 2 above)
Comment (samcgardner) [reply helm/charts/kratos/values.yaml:271]: That seems reasonable. I think Keto uses 100 as its maximum, so you might want to look at changing that as well.

**Comment by samcgardner** [comment]
We use ArgoCD and its default Helm controller. I'm not entirely sure _why_ this works. Now you mention it, it does seem logical for the controller to try and prune the extra field. Perhaps ArgoCD specifically ignores added replica fields for this exact reason?

**Comment by piotrmsc** [comment]
yeah, exactly, I had the same thought that actually it is still the same case. I will take a look a bit deeper to fully understand the behavior

**Comment by piotrmsc** [comment]
for future reference to reproduce the issue easily locally do: 
```
k3d create cluster test

argocd app create hydra --repo https://github.com/ory/k8s.git --path helm/charts/hydra  --dest-server https://kubernetes.default.svc --dest-namespace default --helm-set hydra.config.secrets.system={SOME_SECRET} --helm-set hydra.config.dsn=memory --helm-set hydra.config.urls.self.issuer=https://my-hydra/ --helm-set hydra.config.urls.login=https://my-idp/login --helm-set maester.enabled=false --helm-set image.tag=v1.10.5-sqlite --helm-set deployment.autoscaling.enabled=true --helm-set deployment.autoscaling.minReplicas=2 --helm-set deployment.autoscaling.maxReplicas=3 --helm-set deployment.autoscaling.targetCPUUtilizationPercentage=2 --helm-set deployment.resources.requests.cpu=10m --sync-policy auto --self-heal

```

**Comment by piotrmsc** [comment]
@samcgardner I run into this issue https://github.com/argoproj/argo-cd/issues/1603 where exactly same behavior is described.

However, I think it should be possible to define which fields should be ingored from sync decision in argocd like it's mentioned here https://argoproj.github.io/argo-cd/user-guide/diffing/#application-level-configuration

**Comment by piotrmsc** [comment]
but looks like ignoring replicas this way will not always work https://github.com/argoproj/argo-cd/issues/2913

**Comment by samcgardner** [comment]
Yes, I think this approach would probably work as well. I'm not sure the developer ergonomics are as good, or that it would transfer well to other projects like Flux, so personally I'd prefer to offer this solution in the chart as well since it's fairly standard and harmless

**Comment by piotrmsc** [comment]
I want to understand the behavior more precisely as it's pretty much the same for me in both scenarios TBH.
On ArgoCD slack I received confirmation that ignoring field while diffing is a way to go but no explanation why it behaves differently when replicas are not set in the manifest and then added by HPA.

**Review by piotrmsc — APPROVED** [review]
But in general LGTM! Thank you for the contribution 🎉

**Pull request merged**